src/Data/String/CodePoints.purs

```diff
@@ -0,0 +1,273 @@
+  _ -> Nothing
+
+
+-- | Returns the number of code points in the leading sequence of code points
+-- | which all match the given predicate. Operates in constant space and in
+-- | time linear to the length of the given string.
+count :: (CodePoint -> Boolean) -> String -> Int
+count = _count isLead isTrail unsurrogate
+
+foreign import _count
+  :: (Int -> Boolean)
+  -> (Int -> Boolean)
+  -> (Int -> Int -> CodePoint)
+  -> (CodePoint -> Boolean)
+  -> String
+  -> Int
+
+
+-- | Drops the given number of code points from the beginning of the given
+-- | string. If the string does not have that many code points, returns the
+-- | empty string. Operates in space and time linear to the length of the given
+-- | string.
+drop :: Int -> String -> String
+drop n s = String.drop (String.length (take n s)) s
+
+
+-- | Drops the leading sequence of code points which all match the given
+-- | predicate from the given string. Operates in space and time linear to the
+-- | length of the given string.
+dropWhile :: (CodePoint -> Boolean) -> String -> String
+dropWhile p s = drop (count p s) s
+
+
+-- | Creates a string from an array of code points. Operates in space and time
+-- | linear to the length of the given array.
+fromCodePointArray :: Array CodePoint -> String
+fromCodePointArray = _fromCodePointArray singletonFallback
+
+foreign import _fromCodePointArray
+  :: (CodePoint -> String)
+  -> Array CodePoint
+  -> String
+
+-- | Returns the number of code points preceding the first match of the given
+-- | pattern in the given string. Returns Nothing when no matches are found.
+indexOf :: String.Pattern -> String -> Maybe Int
+indexOf p s = (\i -> length (String.take i s)) <$> String.indexOf p s
+
+
+-- | Returns the number of code points preceding the first match of the given
+-- | pattern in the given string. Pattern matches preceding the given index
+-- | will be ignored. Returns Nothing when no matches are found.
+indexOf' :: String.Pattern -> Int -> String -> Maybe Int
+indexOf' p i s =
+  let s' = drop i s in
+  (\k -> i + length (String.take k s')) <$> String.indexOf p s'
+
+
+-- | Returns the number of code points preceding the last match of the given
+-- | pattern in the given string. Returns Nothing when no matches are found.
+lastIndexOf :: String.Pattern -> String -> Maybe Int
+lastIndexOf p s = (\i -> length (String.take i s)) <$> String.lastIndexOf p s
+
+
+-- | Returns the number of code points preceding the first match of the given
+-- | pattern in the given string. Pattern matches following the given index
+-- | will be ignored. Returns Nothing when no matches are found.
+lastIndexOf' :: String.Pattern -> Int -> String -> Maybe Int
+lastIndexOf' p i s =
+  let i' = String.length (take i s) in
+  (\k -> length (String.take k s)) <$> String.lastIndexOf' p i' s
+
+
+-- | Returns the number of code points in the given string. Operates in
+-- | constant space and time linear to the length of the string.
```

Comment:
Do you think it might be slightly clearer to say "Operates in constant space and *in* time linear to the length of the string"?